### PR TITLE
ekf2: move estimator_status test ratios to filtered values

### PR DESCRIPTION
**Pull request by dagar**

### CODE REVIEW

File: src/modules/ekf2/EKF2.cpp [hunk 1]
```diff
@@ -1802,15 +1802,14 @@ void EKF2::PublishStatus(const hrt_abstime &timestamp)
...
-	_ekf.get_innovation_test_status(innov_check_flags_temp, status.mag_test_ratio,
-					status.vel_test_ratio, status.pos_test_ratio,
-					status.hgt_test_ratio, status.tas_test_ratio,
-					status.hagl_test_ratio, status.beta_test_ratio);
-
-	// Bit mismatch between ecl and Firmware, combine the 2 first bits to preserve msg definition
-	// TODO: legacy use only, those flags are also in estimator_status_flags
-	status.innovation_check_flags = (innov_check_flags_temp >> 1) | (innov_check_flags_temp & 0x1);
+	status.mag_test_ratio = _ekf.getHeadingInnovationTestRatio();
+	status.vel_test_ratio = math::max(_ekf.getHorizontalVelocityInnovationTestRatio(),
```
Comment (dagar) [inline src/modules/ekf2/EKF2.cpp:1806]: This might be problematic if there's no vertical velocity.

**Comment by dagar** [comment]
This also adds aux global position to the position test ratio.

File: msg/EstimatorStatus.msg [hunk 2]
```diff
@@ -69,19 +69,6 @@ uint32 filter_fault_flags	# Bitmask to indicate EKF internal faults
...
-# 4 - true if the Y magnetometer observation has been rejected
-# 5 - true if the Z magnetometer observation has been rejected
-# 6 - true if the yaw observation has been rejected
-# 7 - true if the airspeed observation has been rejected
-# 8 - true if the synthetic sideslip observation has been rejected
-# 9 - true if the height above ground observation has been rejected
-# 10 - true if the X optical flow observation has been rejected
-# 11 - true if the Y optical flow observation has been rejected
 
 float32 mag_test_ratio # ratio of the largest magnetometer innovation component to the innovation test limit
```
Comment (haumarco) [inline msg/EstimatorStatus.msg:73]: Now we can change all of them to
```suggestion
float32 mag_test_ratio # lpf ratio of the largest magnetometer innovation component to the innovation test limit
```
right?

File: src/modules/ekf2/EKF2.cpp (hunk 1 above)
Comment (haumarco) [reply src/modules/ekf2/EKF2.cpp:1806]: But it's basically the same as it used to be, just filtered?

Comment (dagar) [reply src/modules/ekf2/EKF2.cpp:1806]: Now if a vertical velocity isn't available getVerticalVelocityInnovationTestRatio() will return INFINITY and that will be the max for status.vel_test_ratio.

Comment (haumarco) [reply src/modules/ekf2/EKF2.cpp:1806]: is there a reason why you split horizontal and vertical up into 2 functions?

File: msg/EstimatorStatus.msg (hunk 2 above)
Comment (dagar) [reply msg/EstimatorStatus.msg]: Done.

File: src/modules/ekf2/EKF2.cpp (hunk 1 above)
Comment (dagar) [reply src/modules/ekf2/EKF2.cpp]: In anticipation of replacing the preflight checker with the filtered test ratios as well. :smile: 

image

(threshold TBD)

**Comment by dagar** [comment]
Actually, this is making some existing holes/inconsistencies a bit more apparent. 

For example currently in main the `mag_test_ratio` will default to 0, so if you don't have a heading source (mag, gps_yaw) you'll still slip through and pass.

There are other cases like only have a position aid source, but no velocity that should be considered.

File: src/modules/ekf2/EKF2.cpp (hunk 1 above)
Comment (haumarco) [reply src/modules/ekf2/EKF2.cpp]: So you want to keep it like that or handle the case?

File: src/modules/ekf2/EKF/ekf_helper.cpp [hunk 3]
```diff
@@ -301,131 +301,224 @@ void Ekf::resetAccelBias()
...
 #endif // CONFIG_EKF2_EXTERNAL_VISION
 
-	// return the largest velocity and position innovation test ratio
-	vel = NAN;
-	pos = NAN;
+	if (test_ratio >= 0.f && test_ratio <= 1.f) {
+		return sqrtf(test_ratio);
+	}
+
+	return INFINITY;
```
Comment (haumarco) [inline src/modules/ekf2/EKF/ekf_helper.cpp:333]: Why do you want to return infinity if the test ratio is above 1? If the ratio is valid or not will be decided with a gate at a later point.

Comment (dagar) [reply src/modules/ekf2/EKF/ekf_helper.cpp:333]: Right, that doesn't make sense, I think I was being paranoid in a previous iteration with the test ratios initialized to NAN.

**Comment by dagar** [comment]
In the interest of getting this in I'll update this to be consistent with the current behaviour and then propose more of a clean up (possibly eliminating COM_ARM_EKF_YAW, COM_ARM_EKF_VEL, etc) as a follow up.

This means preserving arguably wrong things like mag_test_ratio = 0 if there's no heading source, etc.

**Review by haumarco — APPROVED** [review]

**Comment by dagar** [comment]
I found one minor issue with the mag_test_ratio initially failing due to not resetting the filtered test ratio on mag init (if yaw aligning and resetting heading). I pushed a fix.

image

**Pull request merged**